### ISSUE TITLE
"Failed building wheel for gvar" when installing gvar without numpy (in cache)

### ISSUE THREAD

**Issue by ckoerber** [gplepage/gvar#16]
Hello @gplepage,

First of all, I'd like to thank you for providing and maintaining `gvar` and `lsqfit`. Both are a tremendous help in our analysis projects.

The issue I am finding is an edge case and does not prohibit the installation of `gvar`; however, I believe a fix may speed up installing `gvar` from scratch (by (re-)using wheels).

## The issue

When creating a new virtual environment or installing `gvar` from scratch, I encounter a
```
ModuleNotFoundError: No module named 'numpy'
```
from
```
setup.py", line 32, in build_extensions
```
which causes the building of the [wheel](https://pip.pypa.io/en/stable/cli/pip_wheel/) to fail. However, the install still goes through on most of the systems/versions I have tried (only a few scenarios stop at this step). _As far as I understand it, regarding #2 and #4, you have worked out a way to make it work by delaying the `numpy` import and installing it on the fly._

The steps to reproduce this error is to ensure `numpy` and `gvar` are not installed (i.e., in my case, a new virtual environment) and running
```bash
pip install --no-cache-dir gvar
```

on
```
pip --version 21.3.1
python --version 3.8.10
```

## Possible solution

As far as I understand it, this issue is caused when `numpy` is not present at the build time of `gvar`. Unfortunately, even if you specify `numpy` in `setup.py` `install_requires`, it is not installed before the building starts.

Fortunately, with [PEP 517](https://www.python.org/dev/peps/pep-0517#build-requirements), setup tools, which manages the install in `setup.py`, [now allows to specify dependencies needed at build time](https://setuptools.pypa.io/en/latest/userguide/quickstart.html?highlight=pyproject). This is done by adding a `pyproject.toml` to the root of the directory. In the case of `gvar`, I believe the following content solves the issue I have described above
```
[build-system]
requires = [
    "setuptools >= 40.9.0",
    "wheel",
    "numpy",
    "scipy",
]
build-backend = "setuptools.build_meta"
```
_Generally, to allow for different building tools, it seems to be the trend to go away from "dynamic" setup scripts and make them static. I.e., replacing `setup.py`s with `setup.cfg`s._

I have created a [local fork](https://github.com/ckoerber/gvar) and set up different GitHub actions which try to install `gvar` from scratch on different operating systems and using different distributions.

The ones [with the `pyproject.toml` (see the "install module" tab)](https://github.com/ckoerber/gvar/runs/3660731007?check_suite_focus=true) succeed in building the wheel while the ones  [without the `pyproject.toml`](https://github.com/ckoerber/gvar/runs/4116601540?check_suite_focus=true) fall back to the legacy install, encounter the missing `numpy` Error and proceed without building the wheel.

I believe adding this `.toml` file may also simplify the current `setup.py` file, but I did not look much further into it.

If this is something you are interested in, I am happy to help out and provide more information or create a PR.

Best regards,

Christopher

PS:
* The GitHub actions could also be configured to run the unit tests for PRs or pushes automatically
* I am also happy to look into whether it is possible to convert the `setup.py` into a `setup.cfg`
* This might also apply to `lsqfit`

**Comment by gplepage** [comment]
Thanks so much for taking the time to tell me about this, Christopher. I hadn't noticed that Python had made such significant improvements to its build tools. I need to play with them a little but I am pretty sure this will simplify the ``setup.py`` files, which would be great, as well as fixing the problem you point out.

Peter

**Comment by Gattocrucco** [comment]
Installs are failing in my CI right now due to this problem. I guess that something changed in the Python toolchain and now the current workaround is broken (it previously worked).

**Comment by gplepage** [comment]
Thanks for pointing out the problem (again). This needs to be fixed and I will get to it in about a week.

**Comment by gplepage** [comment]
Finally fixed this, in version 11.11.3. Thanks again, Christopher, for pointing out the problem and the solution.
